### ISSUE TITLE
Selectively Ignore Some Deprovisioning Errors

### ISSUE THREAD

**Issue by spjmurray** [kubernetes-sigs/cluster-api-provider-openstack#1510]
/kind feature

**Describe the solution you'd like**

This is related to https://github.com/argoproj/argo-cd/issues/12943 and it summarizes one facet of the problem, however...

tl;dr CAPI has some pretty strict implicit orderings on deletion, that are not well suited to CD.  For example, deleting infrastructure and machines at the same time may lead to machines not being deprovisioned cleanly due to their related load balancer being deleted first.  What I would like to propose is that during machine deletion, if the load balancer isn't there, just get on with your life, you can imply the load balancer member has been deleted, don't sit in an infinite error state and require manual clean up.

**Anything else you would like to add:**

Yes, I know we should delete the Cluster resource and let nature take its course, however the CD tooling just isn't up to the job yet, and this may, just may, make life easier for the general CAPI crowd.  Worth a punt!

**Comment by mdbooth** [comment]
Do you have an example error? It looks like we should already ignore this: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/af3d78afbf92308a9ca204c1235137ac16dbc071/pkg/cloud/services/loadbalancer/loadbalancer.go#L501-L510

What actually happens?

**Comment by spjmurray** [comment]
Lemme consult grafana... bbiab

**Comment by spjmurray** [comment]
Hah, yeah, your encyclopedic knowledge of the code wins again Matt, excellent support sir!

![Screenshot from 2023-03-21 09-56-09](https://user-images.githubusercontent.com/5549946/226572457-65963f97-76ac-4a43-bbc4-f3068c5d6eea.png)

Yeah so that was a bit of a red herring, it seems the real problem is apparently the OSC (probably?) going bye bye before the machine can be offed.

You'd be well within your rights to tell me it's Argo's problem again 😸   I think this is a common problem with CD just not being mature enough to handle CAPI's deletion semantics, and also provision at the same time.

**Comment by mdbooth** [comment]
Yep, looks like OSCluster is deleted before the OSMachine.

I agree with you on this, btw: the api gives us the tools to delete these things in the correct order under all circumstances, so it would be nice to use them. My only concern is that if we implement this in CAPO but CAPI doesn't share this goal, have we actually solved any practical use case? i.e. Is deleting everything from Argo still going to break your cluster, just in CAPI components instead of CAPO components?

If we could come up with something simple and idiomatic to ensure correct deletion ordering in all circumstances, I think I'd be in favour of including it anyway.

**Comment by spjmurray** [comment]
Right then, seems the correct (for now) thing is to fix argo, which I am loathe to do as fixes take a while, still waiting on one that allows me to delete machine deployments...  So I guess if we provision `secret > infra templates etc > MDs and CL`  then when it comes to deprovisioning, MDs and CL will get killed first, which is fine as CAPI implictly does machines first anyway.  Once that wave is done it's free to kill off infra templates (cluster deprovision would have already done this), at which point it's just a case of fixing the Argo deadlock I linked to.

Close this issue if you like, I know what must be done!
